### PR TITLE
ceph: fixed #4717 Ceph external cluster, repeatedly saving mon endpoi…

### LINKED ISSUE #4717: Ceph external cluster, repeatedly saving mon endpoints to config
**Is this a bug report or feature request?**

* Bug Report

  
**Bug Report**

Rook rewrites configs for external ceph cluster repeatedly every few seconds.

**Expected behavior:**

Only rewrite configs when cluster changes.

**Deviation from expected behavior:**
```
2020-01-17 14:03:25.162995 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:03:25.163036 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:03:25.163052 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:03:25.163064 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:03:25.173956 I | op-mon: saved mon endpoints to config map map[data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789 maxMonId:-1 mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.5:6789","10.1.1.6:6789","10.1.1.7:6789"]}]]
2020-01-17 14:03:25.180472 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:03:25.180747 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-17 14:04:11.090181 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:04:11.090223 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:04:11.090238 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:04:11.090249 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:04:11.100412 I | op-mon: saved mon endpoints to config map map[maxMonId:-1 mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.5:6789","10.1.1.6:6789","10.1.1.7:6789"]}] data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789]
2020-01-17 14:04:11.106600 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:04:11.106910 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-17 14:04:56.987005 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:04:56.987059 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:04:56.987075 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:04:56.987087 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:04:56.997593 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.5:6789","10.1.1.6:6789","10.1.1.7:6789"]}] data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789 maxMonId:-1 mapping:{"node":{}}]
2020-01-17 14:04:57.003829 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:04:57.004133 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-17 14:05:42.858947 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:05:42.859009 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:05:42.859028 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:05:42.859044 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:05:42.871217 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.5:6789","10.1.1.6:6789","10.1.1.7:6789"]}] data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789 maxMonId:-1 mapping:{"node":{}}]
2020-01-17 14:05:42.877706 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:05:42.877999 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
2020-01-17 14:06:28.772799 I | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-17 14:06:28.772840 I | op-mon: new external mon ceph5 found: 10.1.1.5:6789, adding it
2020-01-17 14:06:28.772855 I | op-mon: new external mon ceph6 found: 10.1.1.6:6789, adding it
2020-01-17 14:06:28.772868 I | op-mon: new external mon ceph7 found: 10.1.1.7:6789, adding it
2020-01-17 14:06:28.784259 I | op-mon: saved mon endpoints to config map map[mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.1.1.7:6789","10.1.1.5:6789","10.1.1.6:6789"]}] data:ceph5=10.1.1.5:6789,ceph6=10.1.1.6:6789,ceph7=10.1.1.7:6789 maxMonId:-1]
2020-01-17 14:06:28.789822 I | cephconfig: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-01-17 14:06:28.790111 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph
```

**How to reproduce it (minimal and precise):**

Use rook 1.2.1 with a external cluster with 3 mons.
```
apiVersion: v1
kind: Namespace
metadata:
  name: {{ rook_namespace }}
---
apiVersion: v1
kind: Secret
metadata:
  name: rook-ceph-mon
  namespace: {{ rook_namespace }}
type: Opaque
data:
  cluster-name: {{ rook_clustername | b64encode }}
  fsid: {{ rook_external_fsid | b64encode }}
  admin-secret: {{ rook_external_admin_secret | b64encode }}
  mon-secret: {{ rook_external_mon_secret | b64encode }}
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: rook-ceph-mon-endpoints
  namespace: {{ rook_namespace }}
data:
  data: "{{ rook_external_ceph_mon_data }}"
  mapping: "{}"
  maxMonId: "2"
---
apiVersion: ceph.rook.io/v1
kind: CephCluster
metadata:
  name: {{ rook_clustername }}
  namespace: {{ rook_namespace }}
spec:
  external:
    enable: true
  dataDirHostPath: /var/lib/rook
  # providing an image is optional, do this if you want to create other CRs (rgw, mds, nfs)
  cephVersion:
    image: ceph/ceph:v14.2.6 # MUST match external cluster version
```
**Environment**:
* OS: CentOS 7
* Kernel : 3.10.0-1062.9.1.el7.x86_64
* Cloud provider or hardware configuration: Bare Metal
* Rook version: v1.2.1
* Storage backend version: ceph 14.2.6
* Kubernetes version: v1.17.0
* Kubernetes cluster type: Bare Metal
* Storage backend status: HEALTH_OK

### PR DESCRIPTION
**Pull request by eliaswimmer**
…nts to config

**Description of your changes:**

* Saving old ClusterInfo.Monitors
* Compare old and new ClusterInfo.Monitors

**Which issue is resolved by this Pull Request:**
Resolves #4717 

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

### CODE REVIEW

**Comment by eliaswimmer** [comment]
The test failing in jenkins is from a different package (TestCassandraSuite).

**Comment by rajatsingh25aug** [comment]
We know that, there's an older version of CI running due to which I think
it's failing.
Do a make test locally on your terminal and let me know if there are any
tests failing?.

**Comment by rajatsingh25aug** [comment]
We know that there's an older version of K8 running on the CI which I think has to be different from the k8 version running on your system locally, due to which I think the tests it's failing.
Do a `make test` locally on your terminal and let me know if there are any tests failing for you?.
Thanks

File: pkg/operator/ceph/cluster/mon/health.go [hunk 1]
```diff
@@ -323,12 +323,13 @@ func (c *Cluster) handleExternalMonStatus(status client.MonStatusResponse) error
...
-	// clearing the content of clusterinfo monitors, it'll get populated in the next iteration
-	for mon := range c.ClusterInfo.Monitors {
-		delete(c.ClusterInfo.Monitors, mon)
+	oldClusterInfoMonitors := map[string]*cephconfig.MonInfo{}
+	// clearing the content of clusterinfo monitors, later it'll get populated again
+	for monName, mon := range c.ClusterInfo.Monitors {
+		oldClusterInfoMonitors[mon.Name] = mon
+		delete(c.ClusterInfo.Monitors, monName)
 	}
 	logger.Infof("ClusterInfo is now Empty, refilling it from status.MonMap.Mons")
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/health.go]: Come to think of it, this is clearly a Debug level log.

File: pkg/operator/ceph/cluster/mon/health.go [hunk 2]
```diff
@@ -357,21 +358,36 @@ func (c *Cluster) addOrRemoveExternalMonitor(status client.MonStatusResponse) (b
 				monPort := cephutil.GetPortFromEndpoint(endpoint)
 				logger.Infof("new external mon %s found: %s, adding it", mon.Name, endpoint)
 				c.ClusterInfo.Monitors[mon.Name] = cephconfig.NewMonInfo(mon.Name, monIP, monPort)
-				changed = true
+			} else {
+				logger.Debugf("mon %s is not in quorum and not in ClusterInfo", mon.Name)
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/health.go]:
```suggestion
				logger.Debugf("mon %q is not in quorum and not in ClusterInfo", mon.Name)
```

File: pkg/operator/ceph/cluster/mon/health.go [hunk 3]
```diff
@@ -357,21 +358,36 @@ func (c *Cluster) addOrRemoveExternalMonitor(status client.MonStatusResponse) (b
...
+				// this mon was in clusterInfo but is not part of the quorum anymore
+				// thus don't add it again to ClusterInfo
 				logger.Infof("monitor %s is not part of the external cluster monitor quorum, removing it", mon.Name)
-				delete(c.ClusterInfo.Monitors, mon.Name)
+				changed = true
+			} else {
+				// this mon was in clusterInfo and is still in the quorum
+				// add it again
+				c.ClusterInfo.Monitors[mon.Name] = oldClusterInfoMonitors[mon.Name]
+				logger.Debugf("everything is fine mon %s in the clusterInfo and its quorum status is %v", mon.Name, inQuorum)
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/health.go]:
```suggestion
				logger.Debugf("everything is fine mon %q in the clusterInfo and its quorum status is %v", mon.Name, inQuorum)
```

File: pkg/operator/ceph/cluster/mon/health.go [hunk 4]
```diff
@@ -323,12 +323,13 @@ func (c *Cluster) handleExternalMonStatus(status client.MonStatusResponse) error
 
 func (c *Cluster) addOrRemoveExternalMonitor(status client.MonStatusResponse) (bool, error) {
 	var changed bool
-	// clearing the content of clusterinfo monitors, it'll get populated in the next iteration
-	for mon := range c.ClusterInfo.Monitors {
-		delete(c.ClusterInfo.Monitors, mon)
+	oldClusterInfoMonitors := map[string]*cephconfig.MonInfo{}
+	// clearing the content of clusterinfo monitors, later it'll get populated again
+	for monName, mon := range c.ClusterInfo.Monitors {
+		oldClusterInfoMonitors[mon.Name] = mon
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/health.go]: Please add a comment like: "populate with the current clusterInfo".

File: pkg/operator/ceph/cluster/mon/health.go [hunk 5]
```diff
@@ -357,21 +358,36 @@ func (c *Cluster) addOrRemoveExternalMonitor(status client.MonStatusResponse) (b
...
+				c.ClusterInfo.Monitors[mon.Name] = oldClusterInfoMonitors[mon.Name]
+				logger.Debugf("everything is fine mon %s in the clusterInfo and its quorum status is %v", mon.Name, inQuorum)
+			}
+		}
+	}
+	// compare old clusterInfo with new ClusterInfo
+	if len(oldClusterInfoMonitors) != len(c.ClusterInfo.Monitors) {
+		changed = true
+	} else {
+		// check for similarity
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/health.go]: Why is this necessary?
If we know it's true then let's just return true.

File: pkg/operator/ceph/cluster/mon/health.go [hunk 6]
```diff
@@ -357,21 +358,36 @@ func (c *Cluster) addOrRemoveExternalMonitor(status client.MonStatusResponse) (b
 				monPort := cephutil.GetPortFromEndpoint(endpoint)
 				logger.Infof("new external mon %s found: %s, adding it", mon.Name, endpoint)
 				c.ClusterInfo.Monitors[mon.Name] = cephconfig.NewMonInfo(mon.Name, monIP, monPort)
-				changed = true
+			} else {
+				logger.Debugf("mon %s is not in quorum and not in ClusterInfo", mon.Name)
 			}
-			logger.Debugf("mon %s is not in quorum and not in ClusterInfo", mon.Name)
+			changed = true
```
Comment (leseb) [inline pkg/operator/ceph/cluster/mon/health.go]: Now it feels strange to return True, it seems like it's always True since we re-populate clusterInfo all the time.
If that's the case, we should probably always call `c.saveMonConfig()` after `c.addOrRemoveExternalMonitor(status)` is called.

File: pkg/operator/ceph/cluster/mon/health.go (hunk 5 above)
Comment (eliaswimmer) [reply pkg/operator/ceph/cluster/mon/health.go]: Sorry, but that is exactly the problem. If the new ClusterInfo ist the same as the old one, false should be returned.
All I do is a set compare, if the set differs I return true otherwise false.

File: pkg/operator/ceph/cluster/mon/health.go (hunk 1 above)
Comment (eliaswimmer) [reply pkg/operator/ceph/cluster/mon/health.go]: It is fore sure.

File: pkg/operator/ceph/cluster/mon/health.go (hunk 2 above)
Comment (eliaswimmer) [reply pkg/operator/ceph/cluster/mon/health.go]: Have to look up was %q is for.

File: pkg/operator/ceph/cluster/mon/health.go (hunk 6 above)
Comment (eliaswimmer) [reply pkg/operator/ceph/cluster/mon/health.go]: No really, only if the mon from status is not found in oldClusterInfoMonitors.
If the monitor is found in oldClusterinfoMonitors, it will added to c.ClusterInfo.Monitors again

**Comment by eliaswimmer** [comment]
> Overall I'm not about the change, it looks somehow ok but introduces more changes and the more I think about this the more I think we should simply revert ***[65fb568](https://github.com/rook/rook/commit/65fb5684f28854def6ac719ef0e3e079efa76a87), as the current code is as complex as the initial one.
> 
> Let's address the current issues and decide soon after.
> Thanks for the PR @eliaswimmer

I think this is buggy too! Initially with a prior version of rook I had the problem that suddenly all mons disappeared from the config. After upgrading I got the reported error.

The prior version newer adds the mons from status. Thus mons can only be removed.

File: pkg/operator/ceph/cluster/mon/health.go (hunk 2 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/mon/health.go]: It's just double quoting the string.

File: pkg/operator/ceph/cluster/mon/health.go (hunk 5 above)
Comment (leseb) [reply pkg/operator/ceph/cluster/mon/health.go]: Sorry I got confused, the first if just checks the number if mon and the `else` checks if they are actually the same. Looks good!

**Comment by leseb** [comment]
> > Overall I'm not about the change, it looks somehow ok but introduces more changes and the more I think about this the more I think we should simply revert ***[65fb568](https://github.com/rook/rook/commit/65fb5684f28854def6ac719ef0e3e079efa76a87), as the current code is as complex as the initial one.
> > Let's address the current issues and decide soon after.
> > Thanks for the PR @eliaswimmer
> 
> I think this is buggy too! Initially, with a prior version of rook, I had the problem that suddenly all mons disappeared from the config. After upgrading I got the reported error.
> 
> The prior version newer adds the mons from status. Thus mons can only be removed.

I think my comments should be addressed so we can move forward with that. I'll see if I can test your change too.
Thanks.

**Comment by leseb** [comment]
The patch works, see:

```
2020-01-22 15:08:12.883865 D | op-mon: checking health of mons
2020-01-22 15:08:12.883908 D | op-mon: Acquiring lock for mon orchestration
2020-01-22 15:08:12.883913 D | op-mon: Acquired lock for mon orchestration
2020-01-22 15:08:12.883917 D | op-mon: Checking health for mons in cluster. rook-ceph-external
2020-01-22 15:08:12.884033 D | exec: Running command: ceph quorum_status --connect-timeout=15 --cluster=rook-ceph-external --conf=/var/lib/rook/rook-ceph-external/rook-ceph-external.config --keyring=/var/lib/rook/rook-ceph-external/client.admin.keyring --format json --out-file /tmp/040133931
2020-01-22 15:08:13.439933 I | exec: 2020-01-22 15:08:13.037 7f45fdab8700  1 librados: starting msgr at 
2020-01-22 15:08:13.037 7f45fdab8700  1 librados: starting objecter
2020-01-22 15:08:13.037 7f45fdab8700  1 librados: setting wanted keys
2020-01-22 15:08:13.037 7f45fdab8700  1 librados: calling monclient init
2020-01-22 15:08:13.037 7f45fdab8700  1 librados: starting msgr at 
2020-01-22 15:08:13.037 7f45fdab8700  1 librados: starting objecter
2020-01-22 15:08:13.037 7f45fdab8700  1 librados: setting wanted keys
2020-01-22 15:08:13.037 7f45fdab8700  1 librados: calling monclient init
2020-01-22 15:08:13.040 7f45fdab8700  1 librados: init done
2020-01-22 15:08:13.040 7f45fdab8700  1 librados: init done
2020-01-22 15:08:13.423 7f45fdab8700 10 librados: watch_flush enter
2020-01-22 15:08:13.423 7f45fdab8700 10 librados: watch_flush enter
2020-01-22 15:08:13.423 7f45fdab8700 10 librados: watch_flush exit
2020-01-22 15:08:13.423 7f45fdab8700 10 librados: watch_flush exit
2020-01-22 15:08:13.424 7f45fdab8700  1 librados: shutdown
2020-01-22 15:08:13.424 7f45fdab8700  1 librados: shutdown
2020-01-22 15:08:13.441142 D | op-mon: Mon quorum status: {Quorum:[0 1 2] MonMap:{Mons:[{Name:a Rank:0 Address:10.104.191.142:6789/0 PublicAddr:10.104.191.142:6789/0 PublicAddrs:{Addrvec:[{Type:v2 Addr:10.104.191.142:3300 Nonce:0} {Type:v1 Addr:10.104.191.142:6789 Nonce:0}]}} {Name:b Rank:1 Address:10.96.166.69:6789/0 PublicAddr:10.96.166.69:6789/0 PublicAddrs:{Addrvec:[{Type:v2 Addr:10.96.166.69:3300 Nonce:0} {Type:v1 Addr:10.96.166.69:6789 Nonce:0}]}} {Name:c Rank:2 Address:10.104.199.29:6789/0 PublicAddr:10.104.199.29:6789/0 PublicAddrs:{Addrvec:[{Type:v2 Addr:10.104.199.29:3300 Nonce:0} {Type:v1 Addr:10.104.199.29:6789 Nonce:0}]}}]}}
2020-01-22 15:08:13.441187 D | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-22 15:08:13.441195 D | op-mon: mon "a" is in ClusterInfo, let's test if it's in quorum
2020-01-22 15:08:13.441200 D | op-mon: everything is fine mon "a" in the clusterInfo and its quorum status is true
2020-01-22 15:08:13.441207 I | op-mon: new external mon "b" found: 10.96.166.69:6789, adding it
2020-01-22 15:08:13.441213 I | op-mon: new external mon "c" found: 10.104.199.29:6789, adding it
2020-01-22 15:08:13.441223 D | op-mon: ClusterInfo.Monitors is map[a:0xc0008624c0 b:0xc000cba8a0 c:0xc000cba8e0]
2020-01-22 15:08:13.448278 D | op-mon: updating config map rook-ceph-mon-endpoints that already exists
2020-01-22 15:08:13.454528 I | op-mon: saved mon endpoints to config map map[data:a=10.104.191.142:6789,b=10.96.166.69:6789,c=10.104.199.29:6789 maxMonId:-1 mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph-external","monitors":["10.104.191.142:6789","10.96.166.69:6789","10.104.199.29:6789"]}]]
2020-01-22 15:08:13.458416 D | op-config: updating config secret &Secret{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:rook-ceph-config,GenerateName:,Namespace:rook-ceph-external,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[{ceph.rook.io/v1 CephCluster rook-ceph-external b7dd62dd-8e24-4bc7-8982-94728db1b312 <nil> 0xc0013c8ce8}],Finalizers:[],ClusterName:,Initializers:nil,ManagedFields:[],},Data:map[string][]byte{},Type:kubernetes.io/rook,StringData:map[string]string{mon_host: [v2:10.104.191.142:3300,v1:10.104.191.142:6789],[v2:10.96.166.69:3300,v1:10.96.166.69:6789],[v2:10.104.199.29:3300,v1:10.104.199.29:6789],mon_initial_members: a,b,c,},}
2020-01-22 15:08:13.462129 I | cephconfig: writing config file /var/lib/rook/rook-ceph-external/rook-ceph-external.config
2020-01-22 15:08:13.462241 I | cephconfig: generated admin config in /var/lib/rook/rook-ceph-external
```

Then kiled one monitor:

```
2020-01-22 15:11:19.275933 D | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-22 15:11:19.275937 D | op-mon: mon "a" is in ClusterInfo, let's test if it's in quorum
2020-01-22 15:11:19.275940 D | op-mon: everything is fine mon "a" in the clusterInfo and its quorum status is true
2020-01-22 15:11:19.275942 D | op-mon: mon "b" is in ClusterInfo, let's test if it's in quorum
2020-01-22 15:11:19.275944 D | op-mon: everything is fine mon "b" in the clusterInfo and its quorum status is true
2020-01-22 15:11:19.275946 D | op-mon: mon "c" is in ClusterInfo, let's test if it's in quorum
2020-01-22 15:11:19.275949 I | op-mon: monitor "c" is not part of the external cluster monitor quorum, removing it
2020-01-22 15:11:19.275956 D | op-mon: ClusterInfo.Monitors is map[a:0xc0008624c0 b:0xc000cba8a0]
2020-01-22 15:11:19.282078 D | op-mon: updating config map rook-ceph-mon-endpoints that already exists
2020-01-22 15:11:19.284771 I | op-mon: saved mon endpoints to config map map[mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph-external","monitors":["10.104.191.142:6789","10.96.166.69:6789"]}] data:b=10.96.166.69:6789,a=10.104.191.142:6789 maxMonId:-1]
2020-01-22 15:11:19.286771 D | op-config: updating config secret &Secret{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:rook-ceph-config,GenerateName:,Namespace:rook-ceph-external,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[{ceph.rook.io/v1 CephCluster rook-ceph-external b7dd62dd-8e24-4bc7-8982-94728db1b312 <nil> 0xc0013c8ce8}],Finalizers:[],ClusterName:,Initializers:nil,ManagedFields:[],},Data:map[string][]byte{},Type:kubernetes.io/rook,StringData:map[string]string{mon_host: [v2:10.104.191.142:3300,v1:10.104.191.142:6789],[v2:10.96.166.69:3300,v1:10.96.166.69:6789],mon_initial_members: a,b,},}
```

Re-added the mon:

```
2020-01-22 15:12:50.305499 D | op-mon: ClusterInfo is now Empty, refilling it from status.MonMap.Mons
2020-01-22 15:12:50.305505 D | op-mon: mon "a" is in ClusterInfo, let's test if it's in quorum
2020-01-22 15:12:50.305508 D | op-mon: everything is fine mon "a" in the clusterInfo and its quorum status is true
2020-01-22 15:12:50.305510 D | op-mon: mon "b" is in ClusterInfo, let's test if it's in quorum
2020-01-22 15:12:50.305512 D | op-mon: everything is fine mon "b" in the clusterInfo and its quorum status is true
2020-01-22 15:12:50.306005 I | op-mon: new external mon "c" found: 10.104.199.29:6789, adding it
2020-01-22 15:12:50.306022 D | op-mon: ClusterInfo.Monitors is map[c:0xc0009c4480 a:0xc0008624c0 b:0xc000cba8a0]
2020-01-22 15:12:50.314130 D | op-mon: updating config map rook-ceph-mon-endpoints that already exists
2020-01-22 15:12:50.318893 I | op-mon: saved mon endpoints to config map map[data:a=10.104.191.142:6789,b=10.96.166.69:6789,c=10.104.199.29:6789 maxMonId:-1 mapping:{"node":{}} csi-cluster-config-json:[{"clusterID":"rook-ceph-external","monitors":["10.104.191.142:6789","10.96.166.69:6789","10.104.199.29:6789"]}]]
2020-01-22 15:12:50.321518 D | op-config: updating config secret &Secret{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:rook-ceph-config,GenerateName:,Namespace:rook-ceph-external,SelfLink:,UID:,ResourceVersion:,Generation:0,CreationTimestamp:0001-01-01 00:00:00 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[{ceph.rook.io/v1 CephCluster rook-ceph-external b7dd62dd-8e24-4bc7-8982-94728db1b312 <nil> 0xc0013c8ce8}],Finalizers:[],ClusterName:,Initializers:nil,ManagedFields:[],},Data:map[string][]byte{},Type:kubernetes.io/rook,StringData:map[string]string{mon_host: [v2:10.104.191.142:3300,v1:10.104.191.142:6789],[v2:10.96.166.69:3300,v1:10.96.166.69:6789],[v2:10.104.199.29:3300,v1:10.104.199.29:6789],mon_initial_members: a,b,c,},}
```
